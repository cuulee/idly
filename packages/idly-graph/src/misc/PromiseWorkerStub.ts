import { GetActionTypes } from '../operations/operationsTypes';

// TOFIX WorkerGetStateActionsType only does GetState actions, want to add SetState aswell
export class PromiseWorkerStub {
  /* tslint:disable */
  public _workerCb: (message: GetActionTypes) => Promise<any>;
  registerPromiseWorker(workerCb: (message: GetActionTypes) => Promise<any>) {
    this._workerCb = workerCb;
  }
  /* tslint:enable */
  public postMessage(data: {
    readonly request: any;
    readonly type: string;
  }): Promise<any> {
    const t: GetActionTypes = JSON.parse(JSON.stringify(data));
    if (!t.type) {
      throw new Error('need type');
    }
    if (!t.request) {
      throw new Error('need request');
    }
    return this._workerCb(t);
  }
}
