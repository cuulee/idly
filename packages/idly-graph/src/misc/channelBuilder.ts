import { WorkerActionsType } from '../operations/types';

/**
 * A wrapper around `promiseWorker.postMessage` api
 * its sole purpose is to provide typings and easier testing
 * the main <-> worker calls.
 * @param promiseWorker
 */
export function channelBuilder<T extends WorkerActionsType>(promiseWorker: {
  readonly postMessage: any;
}): (type: T['type']) => (request: T['request']) => Promise<string> {
  return type => {
    return request => {
      const toSend = {
        request,
        type,
      };
      return promiseWorker.postMessage(toSend).catch((e: Error) => {
        console.log('Worker Error', e.message);
        return Promise.reject(e.message);
      });
    };
  };
}
