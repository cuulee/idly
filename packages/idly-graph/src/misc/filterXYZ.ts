import area from '@turf/area';
import bboxClip from '@turf/bbox-clip';
import bboxPolygon from '@turf/bbox-polygon';

import { BBox } from 'idly-common/lib/geo/bbox';
import { mercator } from 'idly-common/lib/geo/sphericalMercator';
import { Tile } from 'idly-common/lib/geo/tile';

export function filterXyz(xyzs: Tile[], bbox: BBox, ratio: number): Tile[] {
  const mainPoly = bboxPolygon(bbox);
  return xyzs.filter(tile => {
    const secPoly = bboxPolygon(mercator.bbox(tile.x, tile.y, tile.z));
    const inter = bboxClip(secPoly, bbox);
    return area(inter) / area(secPoly) >= ratio;
  });
}
