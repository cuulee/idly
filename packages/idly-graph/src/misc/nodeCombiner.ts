import { Feature, Point } from '@turf/helpers';
import { Node } from 'idly-common/lib/osm/structures';

export function nodeCombiner(node: Node, existingProps: {}): Feature<Point> {
  return {
    geometry: {
      coordinates: [node.loc.lon, node.loc.lat],
      type: 'Point',
    },
    id: node.id,
    properties: {
      ...existingProps,
      id: node.id,
    },
    type: 'Feature',
  };
}
