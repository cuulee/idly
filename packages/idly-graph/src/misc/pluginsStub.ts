import { EntityTable } from 'idly-common/lib/osm/structures';

export function pluginsStub(
  geometry: any = {
    'osm_basic--geometry': 'line',
  },
): Promise<any> {
  return Promise.resolve({
    workers: [
      {
        pluginName: 'dummy',
        // tslint:disable no-expression-statement
        worker: (entityTable: EntityTable) => {
          const m = new Map();
          entityTable.forEach((v, k) => {
            m.set(k, geometry);
          });
          return m;
        },
      },
    ],
  });
}
