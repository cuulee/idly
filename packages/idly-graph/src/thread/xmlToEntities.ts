import { attributesGen } from 'idly-common/lib/osm/attributesGen';
import { genLngLat } from 'idly-common/lib/osm/genLngLat';
import { nodeFactory } from 'idly-common/lib/osm/nodeFactory';
import { relationFactory } from 'idly-common/lib/osm/relationFactory';
import { relationMemberGen } from 'idly-common/lib/osm/relationMemberGen';
import {
  Entity,
  LngLat,
  Node as OSMNode,
  Relation,
  RelationMember,
  Tags,
  Way,
} from 'idly-common/lib/osm/structures';
import { tagsFactory } from 'idly-common/lib/osm/tagsFactory';
import { wayFactory } from 'idly-common/lib/osm/wayFactory';

/**
 * Converts the osm xml to an array of Entity,
 * This implementation is inspired from the iD
 * editor. Please look into `Entity` for more details.
 * @param xml osm xml document
 */
export function xmlToEntities(xml: Document): Entity[] {
  if (!xml || !xml.childNodes) {
    return [];
  }
  let root = xml.childNodes[2];
  // tslint:disable no-expression-statement
  if (!root) {
    // TOFIX this is only added to let test cases pass
    root = xml.childNodes[0];
  }
  const children = root.childNodes;
  const entities: Entity[] = [];
  const len = children.length;
  for (let i = 0; i < len; i++) {
    const child = children[i];
    if (child.nodeName === 'node') {
      entities.push(nodeData(child));
    } else if (child.nodeName === 'relation') {
      entities.push(relationData(child));
    } else if (child.nodeName === 'way') {
      entities.push(wayData(child));
    }
  }
  // tslint:enable no-expression-statement

  return entities;
}

function getVisible(attrs: NamedNodeMap): boolean {
  return (
    !attrs.getNamedItem('visible') ||
    attrs.getNamedItem('visible').value !== 'false'
  );
}

function getMembers(obj: any): RelationMember[] {
  const elems = obj.getElementsByTagName('member');
  const members: RelationMember[] = [];
  const len = elems.length;
  for (let i = 0; i < len; i++) {
    const attrs = elems[i].attributes;
    // tslint:disable no-expression-statement
    members.push(
      relationMemberGen({
        id:
          attrs.getNamedItem('type').value[0] + attrs.getNamedItem('ref').value,
        role: attrs.getNamedItem('role').value,
        type: attrs.getNamedItem('type').value,
      }),
    );
    // tslint:enable no-expression-statement
  }
  return members;
}

function getNodes(obj: any): string[] {
  const elems = obj.getElementsByTagName('nd');
  const nodes: string[] = [];
  const len = elems.length;
  // tslint:disable no-expression-statement
  for (let i = 0; i < len; i++) {
    nodes.push('n' + elems[i].attributes.getNamedItem('ref').value);
  }
  // tslint:enable no-expression-statement
  return nodes;
}

function getTags(obj: any): Tags {
  const elems = obj.getElementsByTagName('tag');
  const t: any = {};

  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    t[attrs.getNamedItem('k').value] = attrs.getNamedItem('v').value;
  }
  return tagsFactory(t);
}

function getLoc(attrs: NamedNodeMap): LngLat {
  const lon = attrs.getNamedItem('lon') && attrs.getNamedItem('lon').value;
  const lat = attrs.getNamedItem('lat') && attrs.getNamedItem('lat').value;
  return genLngLat([parseFloat(lon), parseFloat(lat)]);
}

function nodeData(obj: Node): OSMNode {
  const attrs = obj.attributes;
  return nodeFactory({
    attributes: attributesGen({
      changeset:
        attrs.getNamedItem('changeset') &&
        attrs.getNamedItem('changeset').value,
      timestamp:
        attrs.getNamedItem('timestamp') &&
        attrs.getNamedItem('timestamp').value,
      uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value,
      user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
      version: attrs.getNamedItem('version').value,
      visible: getVisible(attrs),
    }),
    id: 'n' + attrs.getNamedItem('id').value,
    loc: getLoc(attrs),
    tags: getTags(obj),
  });
}

function relationData(obj: Node): Relation {
  const attrs = obj.attributes;
  return relationFactory({
    attributes: attributesGen({
      changeset:
        attrs.getNamedItem('changeset') &&
        attrs.getNamedItem('changeset').value,
      timestamp:
        attrs.getNamedItem('timestamp') &&
        attrs.getNamedItem('timestamp').value,
      uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value,
      user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
      version: attrs.getNamedItem('version').value,
      visible: getVisible(attrs),
    }),
    id: 'r' + attrs.getNamedItem('id').value,
    members: getMembers(obj),
    tags: getTags(obj),
  });
}

function wayData(obj: Node): Way {
  const attrs = obj.attributes;
  return wayFactory({
    attributes: attributesGen({
      changeset:
        attrs.getNamedItem('changeset') &&
        attrs.getNamedItem('changeset').value,
      timestamp:
        attrs.getNamedItem('timestamp') &&
        attrs.getNamedItem('timestamp').value,
      uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value,
      user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
      version: attrs.getNamedItem('version').value,
      visible: getVisible(attrs),
    }),
    id: 'w' + attrs.getNamedItem('id').value,
    nodes: getNodes(obj),
    tags: getTags(obj),
  });
}
