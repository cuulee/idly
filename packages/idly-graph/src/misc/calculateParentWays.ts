import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import {
  EntityTable,
  EntityType,
  ParentWays,
} from 'idly-common/lib/osm/structures';

export function calculateParentWays(
  entityTable: EntityTable,
  parentWays: ParentWays = ImMap(),
): ParentWays {
  return parentWays.withMutations(p => {
    entityTable.forEach(w => {
      if (!w || w.type !== EntityType.WAY) {
        return;
      }
      w.nodes.forEach(nodeId => {
        /* tslint:disable */
        p.update(nodeId, (s = ImSet()) => s.add(w.id));
        /* tslint:enable */
      });
    });
  });
}
