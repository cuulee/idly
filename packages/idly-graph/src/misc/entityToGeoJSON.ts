import { Feature } from 'idly-common/lib/osm/feature';
import {
  Entity,
  EntityTable,
  EntityType,
} from 'idly-common/lib/osm/structures';

import { nodeCombiner } from './nodeCombiner';
import { wayCombiner } from './wayCombiner';

// tslint:disable:no-expression-statement object-literal-key-quotes
export function entityToGeoJson(
  entityTable: EntityTable,
  computedProps: any,
): Array<Feature<any, any>> {
  const arr: Array<Feature<any, any>> = [];
  entityTable.forEach((entity: Entity, id) => {
    if (entity.type === EntityType.NODE) {
      arr.push(nodeCombiner(entity, computedProps.get(id)));
    } else if (entity.type === EntityType.WAY) {
      arr.push(wayCombiner(entity, entityTable, computedProps.get(id)));
    } else if (entity.type === EntityType.RELATION) {
      // @TOFIX
    }
  });

  return arr;
}
