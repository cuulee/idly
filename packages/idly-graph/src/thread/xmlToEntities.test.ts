import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import { ParentWays } from 'idly-common/lib/osm/structures';
import { wayFactory } from 'idly-common/lib/osm/wayFactory';

import {
  miniWayXML1,
  miniXML1,
  miniXML2,
  miniXML3,
  nodeXML1,
} from '../misc/fixtures';
import { xmlToEntities } from './xmlToEntities';

const wayXML =
  '<?xml version="1.0" encoding="UTF-8"?><osm>' +
  '<way id="1" visible="true" timestamp="2008-01-03T05:24:43Z" version="1" changeset="522559"><nd ref="1"/></way>' +
  '</osm>';

// tslint:disable:no-expression-statement
describe('parsers', () => {
  describe('node', () => {
    it('matches snapshot', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniXML2, 'text/xml');
      expect(xmlToEntities(xml)).toMatchSnapshot();
      expect(
        xmlToEntities(parser.parseFromString(miniXML1, 'text/xml')),
      ).toMatchSnapshot();
    });
    it('matches a bigger example', () => {
      const parser = new DOMParser();
      const XML = nodeXML1;
      const xml = parser.parseFromString(XML, 'text/xml');
      expect(xmlToEntities(xml)).toMatchSnapshot();
    });
  });
  describe('way', () => {
    it('matches snapshot', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniWayXML1, 'text/xml');
      expect(xmlToEntities(xml)).toMatchSnapshot();
    });
  });

  describe('it makes good parentways', () => {
    it('matches snapshot', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniXML3, 'text/xml');
      expect(xmlToEntities(xml)).toMatchSnapshot();
    });
    it('reuses parent Ways and adds stuff to it', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniXML3, 'text/xml');
      expect(xmlToEntities(xml)).toMatchSnapshot();
    });
  });
});
