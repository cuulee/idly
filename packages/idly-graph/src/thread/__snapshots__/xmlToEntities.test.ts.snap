// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsers it makes good parentways matches snapshot 1`] = `
Array [
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n1",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n2",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n3",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w1",
    "nodes": Array [
      "n1",
      "n2",
    ],
    "tags": Object {},
    "type": "way",
  },
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w2",
    "nodes": Array [
      "n1",
      "n3",
    ],
    "tags": Object {},
    "type": "way",
  },
]
`;

exports[`parsers it makes good parentways reuses parent Ways and adds stuff to it 1`] = `
Array [
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n1",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n2",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n3",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w1",
    "nodes": Array [
      "n1",
      "n2",
    ],
    "tags": Object {},
    "type": "way",
  },
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w2",
    "nodes": Array [
      "n1",
      "n3",
    ],
    "tags": Object {},
    "type": "way",
  },
]
`;

exports[`parsers node matches a bigger example 1`] = `
Array [
  Object {
    "attributes": Object {
      "changeset": "138701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n12",
    "loc": Object {
      "lat": 34.5280866,
      "lon": 74.1795357,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782712",
    "loc": Object {
      "lat": 34.5280866,
      "lon": 74.1795357,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782713",
    "loc": Object {
      "lat": 34.528206,
      "lon": 74.1807924,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782714",
    "loc": Object {
      "lat": 34.5283653,
      "lon": 74.1822907,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782715",
    "loc": Object {
      "lat": 34.528425,
      "lon": 74.1815174,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782716",
    "loc": Object {
      "lat": 34.5285445,
      "lon": 74.1788832,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782717",
    "loc": Object {
      "lat": 34.5290024,
      "lon": 74.1788349,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782718",
    "loc": Object {
      "lat": 34.5292015,
      "lon": 74.179584,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782719",
    "loc": Object {
      "lat": 34.529301,
      "lon": 74.1824357,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782720",
    "loc": Object {
      "lat": 34.5294603,
      "lon": 74.1794149,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782721",
    "loc": Object {
      "lat": 34.5296992,
      "lon": 74.1823873,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782722",
    "loc": Object {
      "lat": 34.5297191,
      "lon": 74.1788349,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782723",
    "loc": Object {
      "lat": 34.5300377,
      "lon": 74.1821698,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782724",
    "loc": Object {
      "lat": 34.5303562,
      "lon": 74.1783274,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782725",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782726",
    "loc": Object {
      "lat": 34.5305354,
      "lon": 74.1805024,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782727",
    "loc": Object {
      "lat": 34.5309137,
      "lon": 74.1799224,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782728",
    "loc": Object {
      "lat": 34.5311526,
      "lon": 74.1786899,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782729",
    "loc": Object {
      "lat": 34.5314114,
      "lon": 74.1793182,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w411770875",
    "nodes": Array [
      "n4133782719",
      "n4133782714",
      "n4133782715",
      "n4133782713",
      "n4133782712",
      "n4133782716",
      "n4133782717",
      "n4133782718",
      "n4133782720",
      "n4133782722",
      "n4133782724",
      "n4133782728",
      "n4133782729",
      "n4133782727",
      "n4133782726",
      "n4133782725",
      "n4133782723",
      "n4133782721",
      "n4133782719",
    ],
    "tags": Object {
      "landuse": "residential",
      "place": "village",
      "source": "MSF Survey",
    },
    "type": "way",
  },
]
`;

exports[`parsers node matches snapshot 1`] = `
Array [
  Object {
    "attributes": Object {
      "changeset": "1",
      "timestamp": "2009-03-07T03:26:33Z",
      "uid": null,
      "user": null,
      "version": "1",
      "visible": true,
    },
    "id": "n1",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {},
    "type": "node",
  },
]
`;

exports[`parsers node matches snapshot 2`] = `
Array [
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782712",
    "loc": Object {
      "lat": 34.5280866,
      "lon": 74.1795357,
    },
    "tags": Object {},
    "type": "node",
  },
  Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782713",
    "loc": Object {
      "lat": 34.528206,
      "lon": 74.1807924,
    },
    "tags": Object {},
    "type": "node",
  },
]
`;

exports[`parsers way matches snapshot 1`] = `
Array [
  Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w411770875",
    "nodes": Array [
      "n4133782719",
      "n4133782714",
      "n4133782715",
      "n4133782713",
      "n4133782712",
      "n4133782716",
      "n4133782717",
      "n4133782718",
      "n4133782720",
      "n4133782722",
      "n4133782724",
      "n4133782728",
      "n4133782729",
      "n4133782727",
      "n4133782726",
      "n4133782725",
      "n4133782723",
      "n4133782721",
      "n4133782719",
    ],
    "tags": Object {
      "landuse": "residential",
      "place": "village",
      "source": "MSF Survey",
    },
    "type": "way",
  },
]
`;
